### DISCUSSION TITLE
Is there a way to re-construct CallGraph generated in the past (so it's faster)?

### DISCUSSION THREAD

**Discussion by brenhub24** [soot-oss/SootUp#1017] [Q&A] [answered]
Hi,

First of all thanks so much for the powerful tool. I'm able to use SootUp to do some interesting analysis of a code base.

Problem is, the CallGraph we constructed is a big one. It took about 50 minutes to generate the call graph with the Class Hierarchy Analysis (CHA) algorithm. That means every time we want to tweak the analysis a little bit and rerun it, we need to wait 50 minutes before seeing any result.....

Is there a faster way to construct the CallGraph? Like for example persist the generated CallGraph into a file and reload it?

Thanks!

Bing

**Comment by swissiety** [comment, marked as answer]
Hi,
that seems to be very annoying. Unfortunately not yet, but we are happy if you like to contribute that feature - I could assign you to #1000 :)

Markus

**Reply by brenhub24** [reply]
yes this missing feature seems to be the solution. not sure if I can pick up the ticket but will have a look....

**Comment by JonasKlauke** [comment]
CHA runs quite slow, probably because it causes a lot of Jimple generation, or there are some issues in the implementation. 
RTA could speed up the process. 
Another improvement could be to classify input locations as "libraries". If the call graph algorithm finds such methods, that are marked as libararies, it wont further investiagtes the called methods of these methods. The resulting call graph will be more focused on the actual application, ignoring the giant JRE or other dependencies.

Edit: that are some work arounds to decrease the runtime, but I agree on swissiety to implement a way to save and load call graphs

**Reply by brenhub24** [reply]
Thanks for the hints. However we have done all of these. RTA is much faster but our code base contains lots of "inject interface & call implementation" usages, only CHA traces into the implementations of the interface calls. We have also classified all "others' codes" as libraries. It's a very old code base......

Yeah looks like issue #1000 is the way to do it.... not sure if I can pick it up but will take a look.

Again thanks for the great library, you are awesome!

**Comment by bingtimren** [comment]
@swissiety  instead of persisting call graph, I implemented a graph pruning feature so that I can generate a partial call graph and not going down the path that I'm not interested.

Please review the PR #1025 , thanks!
